In [1]:
%cd ..

D:\SoftUni\Data Science\ProjectV2


### Imports

In [2]:
import pandas as pd

import numpy as np

from src import functions

from functools import reduce

### Metadata

The data about the rates of different mental illnesses includes information about Schizophrenia, Anxiety, Attention Deficit / Hyperactivity, Bipolar and Depressive disorders. It is age - standardized which means that it takes into account the differences in the age structure of each country. [[1]](#references) The rates are provided in the 'Value' column and are measured in prevalent cases per 100 000 people. Thus the values ​​of different countries could be compared unbiased.

The data is saved in five separate datasets, each with 204 observations of 9 features.

The source of data is https://www.healthdata.org/research-analysis/health-risks-issues/mental-health.

### Loading data

In [3]:
anxiety_disorders_data = pd.read_csv('data/mental_disorders_source_datasets/anxiety_disorders.csv')

In [4]:
anxiety_disorders_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 206 entries, 0 to 205
Data columns (total 9 columns):
 #   Column                    Non-Null Count  Dtype  
---  ------                    --------------  -----  
 0   Location                  206 non-null    object 
 1   Year                      204 non-null    float64
 2   Age                       204 non-null    object 
 3   Sex                       204 non-null    object 
 4   Cause of death or injury  204 non-null    object 
 5   Measure                   204 non-null    object 
 6   Value                     204 non-null    float64
 7   Lower bound               204 non-null    float64
 8   Upper bound               204 non-null    float64
dtypes: float64(4), object(5)
memory usage: 14.6+ KB


In [5]:
attention_deficit_hyperactivity_disorders_data = pd.read_csv('data/mental_disorders_source_datasets/attention_deficit_hyperactivity_disorders.csv')

In [6]:
attention_deficit_hyperactivity_disorders_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 206 entries, 0 to 205
Data columns (total 9 columns):
 #   Column                    Non-Null Count  Dtype  
---  ------                    --------------  -----  
 0   Location                  206 non-null    object 
 1   Year                      204 non-null    float64
 2   Age                       204 non-null    object 
 3   Sex                       204 non-null    object 
 4   Cause of death or injury  204 non-null    object 
 5   Measure                   204 non-null    object 
 6   Value                     204 non-null    float64
 7   Lower bound               204 non-null    float64
 8   Upper bound               204 non-null    float64
dtypes: float64(4), object(5)
memory usage: 14.6+ KB


In [7]:
bipolar_disorders_data = pd.read_csv('data/mental_disorders_source_datasets/bipolar_disorders.csv')

In [8]:
bipolar_disorders_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 206 entries, 0 to 205
Data columns (total 9 columns):
 #   Column                    Non-Null Count  Dtype  
---  ------                    --------------  -----  
 0   Location                  206 non-null    object 
 1   Year                      204 non-null    float64
 2   Age                       204 non-null    object 
 3   Sex                       204 non-null    object 
 4   Cause of death or injury  204 non-null    object 
 5   Measure                   204 non-null    object 
 6   Value                     204 non-null    float64
 7   Lower bound               204 non-null    float64
 8   Upper bound               204 non-null    float64
dtypes: float64(4), object(5)
memory usage: 14.6+ KB


In [9]:
depressive_disorders_data = pd.read_csv('data/mental_disorders_source_datasets/depressive_disorders.csv')

In [10]:
depressive_disorders_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 206 entries, 0 to 205
Data columns (total 9 columns):
 #   Column                    Non-Null Count  Dtype  
---  ------                    --------------  -----  
 0   Location                  206 non-null    object 
 1   Year                      204 non-null    float64
 2   Age                       204 non-null    object 
 3   Sex                       204 non-null    object 
 4   Cause of death or injury  204 non-null    object 
 5   Measure                   204 non-null    object 
 6   Value                     204 non-null    float64
 7   Lower bound               204 non-null    float64
 8   Upper bound               204 non-null    float64
dtypes: float64(4), object(5)
memory usage: 14.6+ KB


In [12]:
substance_use_disorders = pd.read_csv('data/mental_disorders_source_datasets/suds.csv')

In [13]:
substance_use_disorders.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 206 entries, 0 to 205
Data columns (total 9 columns):
 #   Column                    Non-Null Count  Dtype  
---  ------                    --------------  -----  
 0   Location                  206 non-null    object 
 1   Year                      204 non-null    float64
 2   Age                       204 non-null    object 
 3   Sex                       204 non-null    object 
 4   Cause of death or injury  204 non-null    object 
 5   Measure                   204 non-null    object 
 6   Value                     204 non-null    float64
 7   Lower bound               204 non-null    float64
 8   Upper bound               204 non-null    float64
dtypes: float64(4), object(5)
memory usage: 14.6+ KB


In [15]:
schizophrenia_data = pd.read_csv('data/mental_disorders_source_datasets/schizophrenia.csv')

In [16]:
schizophrenia_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 206 entries, 0 to 205
Data columns (total 9 columns):
 #   Column                    Non-Null Count  Dtype  
---  ------                    --------------  -----  
 0   Location                  206 non-null    object 
 1   Year                      204 non-null    float64
 2   Age                       204 non-null    object 
 3   Sex                       204 non-null    object 
 4   Cause of death or injury  204 non-null    object 
 5   Measure                   204 non-null    object 
 6   Value                     204 non-null    float64
 7   Lower bound               204 non-null    float64
 8   Upper bound               204 non-null    float64
dtypes: float64(4), object(5)
memory usage: 14.6+ KB


In [17]:
schizophrenia_data

Location    Year  \
0                                          Afghanistan  2021.0   
1                                               Angola  2021.0   
2                                              Albania  2021.0   
3                                              Andorra  2021.0   
4                                 United Arab Emirates  2021.0   
..                                                 ...     ...   
201                                              Egypt  2021.0   
202                                              Sudan  2021.0   
203                                              China  2021.0   
204  Institute for Health Metrics and Evaluation (I...     NaN   
205  For terms and conditions of use, please visit ...     NaN   

                  Age   Sex Cause of death or injury  \
0    Age-standardized  Both            Schizophrenia   
1    Age-standardized  Both            Schizophrenia   
2    Age-standardized  Both            Schizophrenia   
3    Age-standardized  Both            Schizophrenia   
4    Age-standardized  Both            Schizophrenia   
..                ...   ...                      ...   
201  Age-standardized  Both            Schizophrenia   
202  Age-standardized  Both            Schizophrenia   
203  Age-standardized  Both            Schizophrenia   
204               NaN   NaN                      NaN   
205               NaN   NaN                      NaN   

                         Measure       Value  Lower bound  Upper bound  
0    Prevalent cases per 100,000  218.326830   168.877915   278.399258  
1    Prevalent cases per 100,000  223.959914   172.964445   283.560246  
2    Prevalent cases per 100,000  221.339258   173.404374   281.900774  
3    Prevalent cases per 100,000  271.422686   212.911730   343.806662  
4    Prevalent cases per 100,000  265.869204   207.024389   337.079467  
..                           ...         ...          ...          ...  
201  Prevalent cases per 100,000  242.736640   187.979132   305.356991  
202  Prevalent cases per 100,000  231.438860   179.658136   292.506693  
203  Prevalent cases per 100,000  312.364056   271.691880   356.386410  
204                          NaN         NaN          NaN          NaN  
205                          NaN         NaN          NaN          NaN  

[206 rows x 9 columns]

All datasets have the same shape - 206 rows and 9 columns. They don't contain missing values.

### Feature selection

In [18]:
schizophrenia_data.nunique()

Location                    206
Year                          1
Age                           1
Sex                           1
Cause of death or injury      1
Measure                       1
Value                       204
Lower bound                 204
Upper bound                 204
dtype: int64

All observations in the above datasets are age standardized for 2021. The values of mental disorders are measured by a common unit - Prevalent cases per 100 000. So, each of the columns 'Year', 'Age', 'Sex', 'Cause of death and injury' and 'Measure' has a single unique value for all observations. I will drop them as they don't add any useful value. 'Lower bound' and 'Upper bound' columns also will be dropped as they are not features of interest of this project.

In [19]:
columns_to_drop = ['Lower bound', 'Upper bound', 'Year', 'Age', 'Sex', 'Cause of death or injury', 'Measure']

In [20]:
anxiety_disorders_data.drop(columns = columns_to_drop, inplace = True)
attention_deficit_hyperactivity_disorders_data.drop(columns = columns_to_drop, inplace = True)
bipolar_disorders_data.drop(columns = columns_to_drop, inplace = True)
depressive_disorders_data.drop(columns = columns_to_drop, inplace = True)
schizophrenia_data.drop(columns = columns_to_drop, inplace = True)
substance_use_disorders.drop(columns = columns_to_drop, inplace = True)

### Merging the data about mental disorders rates into a common dataset

In [21]:
# The 'Value' columns will be renamed for better description of the features. 'Location' column will be renamed to 'country' for consistency with the other dataset.
anxiety_disorders_data = anxiety_disorders_data.rename(columns = {'Value': 'anxiety_disorders_rate', 'Location': 'country'})
attention_deficit_hyperactivity_disorders_data = attention_deficit_hyperactivity_disorders_data.rename(columns = {'Value': 'adh_disorders_rate', 'Location': 'country'})
bipolar_disorders_data = bipolar_disorders_data.rename(columns = {'Value': 'biploar_disorders_rate', 'Location': 'country'})
depressive_disorders_data = depressive_disorders_data.rename(columns = {'Value': 'depressive_disorders_rate', 'Location': 'country'})
schizophrenia_data = schizophrenia_data.rename(columns = {'Value': 'schizophrenia_rate', 'Location': 'country'})
substance_use_disorders_data = substance_use_disorders.rename(columns = {'Value': 'suds_rate', 'Location': 'country'})

In [23]:
mental_disorders_data = reduce(
    lambda x, y: pd.merge(x, y, on = 'country', how = 'inner'), [
        anxiety_disorders_data, 
        attention_deficit_hyperactivity_disorders_data, 
        bipolar_disorders_data, 
        depressive_disorders_data, 
        schizophrenia_data, 
        substance_use_disorders_data
    ]
)

### Exploration and cleaning of resulted dataset

In [24]:
mental_disorders_data

country  \
0                                          Afghanistan   
1                                               Angola   
2                                              Albania   
3                                              Andorra   
4                                 United Arab Emirates   
..                                                 ...   
201                                              Egypt   
202                                              Sudan   
203                                              China   
204  Institute for Health Metrics and Evaluation (I...   
205  For terms and conditions of use, please visit ...   

     anxiety_disorders_rate  adh_disorders_rate  biploar_disorders_rate  \
0               6036.082637         1005.577242              710.996105   
1               4124.190051          484.173188              501.630997   
2               4620.548604         1014.314099              479.798202   
3               6651.337409         1290.713146              819.692991   
4               5168.099462          447.507189              712.909767   
..                      ...                 ...                     ...   
201             5213.926877         1009.845662              687.032110   
202             5734.418328         1001.996846              714.821440   
203             3481.741558         2183.990788              173.056811   
204                     NaN                 NaN                     NaN   
205                     NaN                 NaN                     NaN   

     depressive_disorders_rate  schizophrenia_rate   suds_rate  
0                  5915.655797          218.326830  412.593458  
1                  6772.060127          223.959914  276.524626  
2                  2977.090487          221.339258  553.140324  
3                  4495.301885          271.422686  789.456777  
4                  4332.425886          265.869204  731.819748  
..                         ...                 ...         ...  
201                4605.408486          242.736640  368.037296  
202                5034.728374          231.438860  378.282560  
203                2875.676654          312.364056  539.805342  
204                        NaN                 NaN         NaN  
205                        NaN                 NaN         NaN  

[206 rows x 7 columns]

In [25]:
mental_disorders_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 206 entries, 0 to 205
Data columns (total 7 columns):
 #   Column                     Non-Null Count  Dtype  
---  ------                     --------------  -----  
 0   country                    206 non-null    object 
 1   anxiety_disorders_rate     204 non-null    float64
 2   adh_disorders_rate         204 non-null    float64
 3   biploar_disorders_rate     204 non-null    float64
 4   depressive_disorders_rate  204 non-null    float64
 5   schizophrenia_rate         204 non-null    float64
 6   suds_rate                  204 non-null    float64
dtypes: float64(6), object(1)
memory usage: 11.4+ KB


In [26]:
mental_disorders_data.duplicated().unique()

array([False])

The resulted dataset has 206 rows and 7 features.

There are no duplicated observations nor missing values.

The last two rows are not observatios but contain information about data provider and terms of use. It is safe to delete them.

In [27]:
mental_disorders_data = mental_disorders_data.iloc[:-2]

In [28]:
mental_disorders_data.tail(2)

country  anxiety_disorders_rate  adh_disorders_rate  \
202   Sudan             5734.418328         1001.996846   
203   China             3481.741558         2183.990788   

     biploar_disorders_rate  depressive_disorders_rate  schizophrenia_rate  \
202              714.821440                5034.728374          231.438860   
203              173.056811                2875.676654          312.364056   

      suds_rate  
202  378.282560  
203  539.805342

In [29]:
# The float values will be rounded for better readability.
mental_disorders_data = mental_disorders_data.round(0)

In [30]:
mental_disorders_data.head(3)

country  anxiety_disorders_rate  adh_disorders_rate  \
0  Afghanistan                  6036.0              1006.0   
1       Angola                  4124.0               484.0   
2      Albania                  4621.0              1014.0   

   biploar_disorders_rate  depressive_disorders_rate  schizophrenia_rate  \
0                   711.0                     5916.0               218.0   
1                   502.0                     6772.0               224.0   
2                   480.0                     2977.0               221.0   

   suds_rate  
0      413.0  
1      277.0  
2      553.0

In [31]:
mental_disorders_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 204 entries, 0 to 203
Data columns (total 7 columns):
 #   Column                     Non-Null Count  Dtype  
---  ------                     --------------  -----  
 0   country                    204 non-null    object 
 1   anxiety_disorders_rate     204 non-null    float64
 2   adh_disorders_rate         204 non-null    float64
 3   biploar_disorders_rate     204 non-null    float64
 4   depressive_disorders_rate  204 non-null    float64
 5   schizophrenia_rate         204 non-null    float64
 6   suds_rate                  204 non-null    float64
dtypes: float64(6), object(1)
memory usage: 11.3+ KB


The dtypes of features in resulted dataframe are appropriate so no need to transform them.

### Saving transformed data
The resulted dataset is saved for further use in the project.

In [32]:
mental_disorders_data.to_csv('data/preprocessed_datasets/mental_disorders_by_country.csv', index = False)

<a id='references'></a>

### References
1. [Age standardization - ourworldindata.org](https://ourworldindata.org/age-standardization)